경북대 BE_박신영 3단계 - 위시 리스트 구현

parknew0 opened the pull request:
안녕하세요, 멘토님. 3단계 위시리스트 기능을 구현했습니다.

## 구현 내용
- 로그인한 사용자를 위한 위시리스트 기능을 구현했습니다.
- 위시리스트 상품 추가 (POST /api/wishes)
- 위시리스트 목록 조회 (GET /api/wishes)
- 위시리스트 상품 삭제 (DELETE /api/wishes/products/{productId})

## 주요 변경 및 리팩토링
컨트롤러에서 인증된 사용자 정보를 받기 위해, 기존 @RequestAttribute 방식을 HandlerMethodArgumentResolver와 커스텀 어노테이션인 @LoginMember를 사용하는 방식으로 리팩토링하여 결합도를 낮추고 가독성을 향상시켰습니다.

## 고민했던 점
위시리스트에 '수량' 개념을 포함할지 고민했으나, '장바구니'와 역할을 구분하기 위해 '찜' 기능에 집중하는 단순한 모델로 구현했습니다.
위시리스트는 장바구니와는 달리, 어떤 상품을 살지 고민하는 과정에서 물건 체크만 하는 행위라 생각해서 수량 개념은 필요없다고 판단했기 때문입니다.

parknew0 committed:
refactor(security): spring-boot-starter-security 의존성 제거 및 리팩토링

parknew0 committed:
refactor(auth): request attribute의 userId를 memberId로 변경

parknew0 committed:
refactor(Member): 생성 로직에 정적 팩토리 메소드 패턴 적용

 - 리팩토링 과정에서 `BeanPropertyRowMapper`가 사용하는 기본 생성자가 없어 로그인 시 객체 매핑에 실패하는 문제가 발생함.프레임워크와의 호환성을 위해 기본 생성자는 public으로 유지

parknew0 committed:
docs(readme): 3단계 위시리스트 기능 목록 추가

parknew0 committed:
docs(readme): 리드미 오타 수정

parknew0 committed:
feat(wish): add wish entity, dto, and schema

- 위시리스트는 '관심 상품'에 가깝다. "나중에 살 수도 있는" 상품들을 모아두는 목록이다. 여기서는 상품의 존재 여부가 중요하지, 수량은 부차적일 것이라 생각했다. 따라서 수량은 DB 설계에 포함하지 않았다.

parknew0 committed:
feat(wish): 위시리스트 상품 추가 API 구현

parknew0 committed:
feat(wish): 위시리스트 목록 조회 API 구현

- `GET /api/wishes` 엔드포인트 추가
- `WishRepository`에 `findByMemberId` 메서드 추가
- 관련 서비스 및 컨트롤러 로직 구현

parknew0 committed:
fix(docs): 위시리스트 목록 조회 API 구현 내역을 리드미에 반영

parknew0 committed:
feat(wish): 위시리스트 상품 삭제 API 구현

- `DELETE /api/wishes/products/{productId}` 엔드포인트 추가
- `Repository`에 `deleteByMemberIdAndProductId` 메서드 추가

parknew0 committed:
refactor(auth): @LoginMember 어노테이션 및 ArgumentResolver 적용

parknew0 committed:
fix(wish): 위시리스트에 없는 상품 삭제 시 404 응답 반환

parknew0 committed:
fix(docs): `@LoginMember` 어노테이션 및 `HandlerMethodArgumentResolver` 구현 내역을 리드미에 반영

rayfault commented on src/main/java/gift/service/WishService.java:
```diff
@@ -0,0 +1,42 @@
+    public List<ProductResponse> getWishes(Long memberId) {
+        List<Wish> wishes = wishRepository.findByMemberId(memberId);
+        return wishes.stream()
+                .map(wish -> productService.findProductById(wish.getProductId()))
```
쿼리를 한 번에 수행하는 것은 어떨까요?
현재 위시리스트에 상품이 20개 있다면, 상품을 20번 조회해와야할텐데 한 번에 상품을 모두 불러올 수는 없을까요?

parknew0 replied:
기존의 코드는 한 번에 하나씩 조회를 해서 네트워크 통신과 데이터베이스에 과부하를 일으킬 수 있었네요,
위시리스트 목록을 가져온 뒤 ID를 수집해서 한 번만에 모든 상품 정보를 가져올 수 있게끔 수정했습니다!
추가적으로 `WishService`가 `ProductService` 대신 `ProductRepository`를 직접 사용하도록 의존성을 변경해서 결합도를 낮추고 책임을 더 명확하게 분리시켰습니다.

rayfault commented on src/main/java/gift/repository/JdbcWishRepository.java:
```diff
@@ -0,0 +1,55 @@
+
+    @Override
+    public List<Wish> findByMemberId(Long memberId) {
+        String sql = "SELECT member_id, product_id FROM wishes WHERE member_id = ?";
```
id 를 가져오지 않는 이유가 있을까요? id 까지 가져온다면 rowMapper 를 따로 구현하지 않아도 될테구요.

parknew0 replied:
이전에는 현재 기능엔 id가 필요 없으니, productId만 가져오도록 커스텀을 했었는데,
표준 도구(BeanPropertyRowMapper)써서 id까지 전부 가져와서 완전한 객체를 만드는 방식으로 개선 했습니다!

rayfault replied:
BeanPropertyRowMapper 가 표준은 아니고, 절대적인 것도 아니지만 직접 작성하는 코드의 양을 줄여주기에 선호하는 편입니다.
대신 조건이 좀 맞아야겠죠 :)

rayfault commented on src/main/java/gift/entity/Member.java:
```diff
@@ -6,18 +6,18 @@ public class Member {
     }
 
-    public Member(Long id, String email, String password) {
+    private Member(Long id, String email, String password) {
```
이 메소드도 static 메소드로 바꾼다면 기본 생성자를 지울 수 있으리라 기대해봅니다.

parknew0 replied:
초기 멘토님의 피드백에 따라, JdbcMemberRepository에 수동으로 구현된 RowMapper를 스프링이 제공하는 BeanPropertyRowMapper로 교체했습니다. 이 과정에서 BeanPropertyRowMapper의 작동 조건(기본 생성자 및 Setter 필요)을 만족시키기 위해 Member 엔티티는 변경 가능한(Mutable) 객체가 되었습니다.

이후 멘토님께서 Member 엔티티를 static으로 변경하라는 피드백을 받고,
Member 엔티티의 기본 생성자와 Setter를 제거하고, private 생성자와 static 팩토리 메소드만으로 객체를 생성하도록 구조를 변경했습니다. 이 설계 변경에 따라 BeanPropertyRowMapper의 사용 조건을 만족시킬 수 없게 되어, JdbcMemberRepository에서는 static 팩토리 메소드를 호출하는 수동 RowMapper를 다시 구현하여 적용했습니다.

이 과정에서 궁금한 점이 있는데,
RowMapper -> BeanPropertyRowMapper -> RowMapper 로 다시 돌아오는 과정을 거치게 되었는데, 멘토님이 의도하신대로 진행되고 있는게 맞을까요??

RowMapper -> BeanPropertyRowMapper 로 변경하라는 피드백에서는, 직접 만들기 보단 SpringFramework 에서 제공하는 RowMapper 구현체를 만드는 편이 편리하기 때문에 사용해보라는 취지에서 말씀하신 것 같고,
 
BeanPropertyRowMapper -> RowMapper 로 변경하라는 피드백에서는, 애플리케이션의 핵심이 되는 객체는 안정성을 최우선으로 고려하여 불변으로 설계하는 편이 더 나아서 다시 되돌리라는 뜻으로 이해했습니다!

rayfault replied:
제가 혼란을 드렸나 보군요.

가장 원했던 것은 아래의 몇 줄을 지우고 BeanPropertyRowMapper 를 사용하는 것이었습니다.

```java
// BeanPropertyRowMapper가 객체를 생성하기 위해선 삭제하면 안됨
public Member() {
}
```

이 과정에서 기본 생성자 + setter 의 조합이 필요하다 말씀하셨습니다.
원했던 것을 풀어보자면,

```java
public Member(Long id, String email, String password) {
...
```

위 생성자를 아래와 같이 정적 팩토리 메소드로 바꾸면서 파라미터를 여러개 받는 생성자를 제거하면
```java
public static Member of(Long id, String email, String password) {
    Member member = new Member();
    member.setId(id);
    member.setEmail(email);
    member.setPassword(password);
    return member;
}
```

Member 클래스에는 선언해 둔 기본 생성자만 남아있기에 이 기본 생성자를 삭제할 수 있게 됩니다.
자바에서는 다른 생성자가 없다면 자동으로 기본 생성자가 생성되죠 :)

그래서 최종적으로 기대한 Member 클래스는 아래와 같습니다.

```java
package gift.entity;

public class Member {

    private Long id;
    private String email;
    private String password;

    public void setId(Long id) {
        this.id = id;
    }

    public void setEmail(String email) {
        this.email = email;
    }

    public void setPassword(String password) {
        this.password = password;
    }

    public Long getId() {
        return id;
    }

    public String getEmail() {
        return email;
    }

    public String getPassword() {
        return password;
    }

    public static Member createNewMember(String email, String password) {
        return of(null, email, password);
    }

    public static Member of(Long id, String email, String password) {
        Member member = new Member();
        member.setId(id);
        member.setEmail(email);
        member.setPassword(password);
        return member;
    }
}
```

* 기본생성자와 Setter 가 존재하고, (기본 생성자는 JAVA 가 자동으로)
* 정적 팩토리 메소드를 통해 새로운 객체를 생성


답변이 되셨을까요? 제가 너무 함축적? 으로 이야기를 해서 많이 고민하셨을것 같네요. 죄송합니다. 🙇🏻‍♂️

rayfault commented on src/main/java/gift/entity/Member.java:
```diff
@@ -6,18 +6,18 @@ public class Member {
     private String email;
     private String password;
 
+    // BeanPropertyRowMapper가 객체를 생성하기 위해선 삭제하면 안됨
```
Repository 동작에 대한 테스트가 있다면 이 주석이 없더라도 테스트를 통해 인지할 수 있지 않을까요?
가능하다면 테스트를 한 번 추가해보면 좋겠습니다. :)

parknew0 replied:
프로젝트 사이즈가 점점 커짐에 따라 테스트 코드의 필요성을 크게 느끼고 있습니다. 다음 과제 단계에서 테스트코드를 추가해보도록 하겠습니다!

rayfault reviewed (changes requested):
전반적으로 깔끔하게 요구사항을 구현해주셨습니다.
큰 군더더기가 없어서 개선에 투자를 해보면 어떨까 싶은데요. 코멘트 한 번 보시고 결정 부탁드립니다.

이번의 코멘트는 필수사항은 아니니 수정여부를 결정해주시고, 코멘트 달아주시면 좋겠습니다.
1차 리뷰 마치겠습니다. 고생하셨습니다~

parknew0 committed:
refactor(wish): 여러번 쿼리를 통해 상품을 하나씩 가져오던 상황을 한 번의 쿼리로 모든 상품 목록 가져오도록 개선, 추가로 `WishService`가 `ProductService` 대신 `ProductRepository`를 직접 사용하도록 의존성을 변경

parknew0 committed:
refactor(wish): WishRepository에 BeanPropertyRowMapper 적용

parknew0 committed:
refactor(member): Member 엔티티를 static 팩토리 메소드를 사용하는 불변 객체로 변경

Member 엔티티의 안정성과 예측 가능성을 높이기 위해, 한번 생성된 후에는 상태가 변하지 않는 불변(immutable) 객체로 리팩토링했습니다.

- Member 엔티티의 기본 생성자와 Setter를 모두 제거하고, 생성자를 private으로 변경하여 객체의 불변성을 보장합니다.
- 객체 생성은 `id` 유무에 따라 `from()`과 `of()` static 팩토리 메소드를 통해서만 이루어지도록 일관성을 확보했습니다.
- 이 설계 변경에 따라, `JdbcMemberRepository`에서 더 이상 `BeanPropertyRowMapper`를 사용할 수 없으므로, 새로운 static 팩토리 메소드를 호출하는 수동 `RowMapper`를 구현하여 적용했습니다.

rayfault reviewed (approved):
코멘트 반영해주신것 확인하였습니다.

어느정도 기대한 대로 수정되었는데요. 작성해주신 코멘트를 보니 제 의도를 파악하느라 고민을 많이 하셨겠다 싶네요. 죄송합니다. 🙇🏻‍♂️ 🙇🏻‍♂️ 저도 좀 더 자세하게 적는 습관을 들여야겠네요.

이번 미션, 단계는 이번 PR 로 마무리짓게 되는데요. 그간 코멘트 반영하느라 고생 많으셨습니다.
저는 덕분에 즐겁게 코드리뷰 하면서 새로운 시각들을 볼 수 있어서 좋았습니다.

리뷰 마치겠습니다. 고생하셨습니다.
